src/mongo/Config.class.php

```diff
@@ -1762,50 +1773,49 @@ protected function getConnectionForDataSource($dataSource)
         {
             $retries = 1;
             $exception = null;
-            $connected = false;
 
             do {
                 try {
-                    $this->connections[$dataSource] = $this->getMongoClient($ds['connection'], $connectionOptions);
-                    // Check if we are connected
-                    $connections = $this->connections[$dataSource]->getConnections();
-                    if (empty($connections) === false) {
-                        $connected = true;
-                    }
-                } catch (\MongoConnectionException $e) {
-                    self::getLogger()->error("MongoConnectionException attempt ".$retries.". Retrying...:" . $e->getMessage());
+                    $connectionString = $ds['connection'] . '?' . http_build_query($connectionOptions);
+                    $this->connections[$dataSource] = $this->getMongoClient($connectionString);
+                    break;
+                } catch (ConnectionTimeoutException $e) {
+                    self::getLogger()->error("ConnectionTimeoutException attempt ".$retries.". Retrying...:" . $e->getMessage());
                     sleep(1);
                     $retries++;
                     $exception = $e;
                 }
 
-            } while ($retries <= self::CONNECTION_RETRIES && $connected === false);
+            } while ($retries <= self::CONNECTION_RETRIES);
 
             if (!isset($this->connections[$dataSource])) {
                 self::getLogger()->error("MongoConnectionException failed after " . $retries . " attempts (MAX:".self::CONNECTION_RETRIES."): " . $e->getMessage());
-                throw new \MongoConnectionException($exception);
+                throw new ConnectionTimeoutException($exception);
```

Comment:
No nasty ping - if the connection object is created successfully (no exception thrown), then i'm breaking on the line afterwards to get out of the do..while loop.